sound/soc/sof/mediatek/mt8195/mt8195.c

```diff
@@ -316,6 +316,20 @@ static int mt8195_dsp_suspend(struct snd_sof_dev *sdev, u32 target_state)
 {
 	struct platform_device *pdev = container_of(sdev->dev, struct platform_device, dev);
 	int ret;
+	u32 reset_sw, dbg_pc;
+
+	/* wait dsp enter idle, timeout is 1 second */
+	ret = snd_sof_dsp_read_poll_timeout(sdev, DSP_REG_BAR,
+					    DSP_RESET_SW, reset_sw,
+					    ((reset_sw & ADSP_PWAIT) == ADSP_PWAIT),
+					    SUSPEND_DSP_IDLE_POLL_INTERVAL_US,
+					    SUSPEND_DSP_IDLE_TIMEOUT_US);
+	if (ret < 0) {
+		dbg_pc = snd_sof_dsp_read(sdev, DSP_REG_BAR, DSP_PDEBUGPC);
+		dev_err(sdev->dev, "dsp not idle : swrest %#x, pc %#x\n",
+			reset_sw, dbg_pc);
+		return ret;
```

Comment:
if you return an error in the suspend path, does it not prevent the entire system from sleeping? is this a desired outcome, or should you reset on resume?

Reply:
May I also please ask what this means "When DSP receive suspend message, DSP firmware will send interrupt to
notify host then DSP will enter idle mode". When does the DSP receive the suspend message? Does the host send it? It doesnt seem to be sent in the the suspend callback.

Reply:
@ranj063  In the sof_suspend function, host will sent ipc with "SOF_IPC_PM_CTX_SAVE" message to DSP for upcoming power down as below 
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/pm.c#L233
```c
	/* notify DSP of upcoming power down */
	if (pm_ops && pm_ops->ctx_save) {
		ret = pm_ops->ctx_save(sdev);
```
DSP will handle this ipc message and set pm_prepare_D3  as 1 as below,
https://github.com/thesofproject/sof/blob/main/src/ipc/ipc3/handler.c#L660
```c
static int ipc_pm_context_save(uint32_t header)
{
...
	ipc_get()->pm_prepare_D3 = 1;
}
```
then DSP will enter wait for interrupt in the ipc_platform_complete_cmd function. Thanks
https://github.com/thesofproject/sof/blob/main/src/drivers/mediatek/mt8195/ipc.c#L84
```c
void ipc_platform_complete_cmd(struct ipc *ipc)
{
	trigger_irq_to_host_rsp();
	while (ipc->pm_prepare_D3) {
		clock_set_freq(CLK_CPU(cpu_get_id()), CLK_SUSPEND_CPU_HZ);
		asm volatile("waiti 15");
	}
}
```

Reply:
@yaochunhung Ahh the CTX_SAVE IPC. So are you saying that there's a new IPC recevied from the DSP after the CTX_SAVE IPC is sent or are you saying that the reply for the CTX_SAVE IPC is the cause for the race?

Reply:
@yaochunhung thanks. 1s timeout during suspend seems excessive.  Have you seen it take that long really for the DSP to idle? Also, I suppose it wont be possible to send the reply after clock_set_free, right?

Reply:
@ranj063  In normal cases, it should be short time for DSP to enter idle. (less than 2ms). You are right that DSP will send reply to host first then do clock_set_freq and enter wait for interrupt (idle) as below, thanks. 
```c
void ipc_platform_complete_cmd(struct ipc *ipc)
{
	trigger_irq_to_host_rsp(); <= DSP send interrupt to notify(reply) host. 
	while (ipc->pm_prepare_D3) {
		clock_set_freq(CLK_CPU(cpu_get_id()), CLK_SUSPEND_CPU_HZ);
		asm volatile("waiti 15");
	}
}
```

Reply:
@ranj063  ipc_platform_complete_cmd is a common api for ipc and trigger_irq_to_host_rsp is always sent.  If your purpose is to reduce time for dsp to enter idle. I think it should be modified as below, but I don't think it will spend much time to set clock_set_freq therefore I set it after trigger_irq_to_host_rsp. thanks
```c
void ipc_platform_complete_cmd(struct ipc *ipc)
{
       if (ipc->pm_prepare_D3)
            clock_set_freq(CLK_CPU(cpu_get_id()), CLK_SUSPEND_CPU_HZ);

        trigger_irq_to_host_rsp(); <= DSP send interrupt to notify(reply) host. 
	while (ipc->pm_prepare_D3)
		asm volatile("waiti 15");
}
```

Reply:
@yaochunhung thanks, I suppose even with this we can never be certain that the DSP is indeed idle. I'm fine with this PR. Thanks!

Reply:
@plbossart  I am also a bit confused about which one is better about  return error to stop sleeping or resume it in resume. I see there is also check dsp power down in the function hda_suspend then return error value. For mt8195, I think it should be fine to reset dsp in resume function to avoid preventing entire system sleeping.  I will modify it. Thanks.    
```c
static int hda_suspend(struct snd_sof_dev *sdev, bool runtime_suspend)
{
...
	/* power down DSP */
	ret = snd_sof_dsp_core_power_down(sdev, chip->host_managed_cores_mask);
	if (ret < 0) {
		dev_err(sdev->dev,
			"error: failed to power down core during suspend\n");
		return ret;
	}
}
```